Add current beatmap info to window title

zihadmahiuddin opened the pull request:
I have tested it on Linux, both on Wayland and X11, using SDL2 and SDL3, and it's working fine for me.
It might need more changes as I'm still not super familiar with the codebase, but I thought I'd create a PR and continue here.

Closes #32188

zihadmahiuddin committed:
Show current beatmap info in window title

zihadmahiuddin committed:
Skip window title update for dummy beatmap

Susko3 commented on osu.Game/OsuGame.cs:
```diff
@@ -828,6 +828,12 @@ private void beatmapChanged(ValueChangedEvent<WorkingBeatmap> beatmap)
             beatmap.NewValue?.BeginAsyncLoad();
+
+            // prevent weird window title saying please load a beatmap
+            if (beatmap.NewValue is null or DummyWorkingBeatmap)
+                Host.Window.Title = Name;
+            else
+                Host.Window.Title = $"{Name} - {beatmap.NewValue.BeatmapInfo.GetDisplayTitleRomanisable(true, false)}";
```
It's safer to schedule setting the window title on the main (input) thread:

```cs
string newTitle = ...;

Host.InputThread.Scheduler.AddOnce(s => Host.Window.Title = s, newTitle);
```

zihadmahiuddin replied:
That makes sense, I forgot to account for that. I'll run the tests before pushing a fix.

zihadmahiuddin replied:
My local tests haven't all finished running yet but they don't fail immediately anymore, so I think the fix is working.

bdach replied:
I'm going to be real here for a second, *should the consumer of the framework care about correct threading* here? I'd say no, I'd say that the framework should automatically schedule correctly. Does it not? Where is this coming from? @smoogipoo am I alone in thinking this way?

zihadmahiuddin replied:
I agree, I assumed that the framework would be able to handle it by itself so I didn't do anything for that manually.

peppy replied:
I'd be fine if this was an update thread requirement, but yeah, secret knowledge that this should be run on the input thread is bad..

zihadmahiuddin replied:
I tested it and looks like it already runs in the desired thread.
I replaced `Host.InputThread.Scheduler.AddOnce` with `Scheduler.AddOnce` to make it run on the update thread, then I set a breakpoint in the setter of `SDL2Window.Title` and checked that [the use of `ScheduleCommand`](https://github.com/ppy/osu-framework/blob/e7f3343f3022dd32f2c8264e7cc421bf5414a4b0/osu.Framework/Platform/SDL2/SDL2Window.cs#L71) makes sure it runs on the desired thread.
[`SDL3Window` does the same](https://github.com/ppy/osu-framework/blob/e7f3343f3022dd32f2c8264e7cc421bf5414a4b0/osu.Framework/Platform/SDL3/SDL3Window.cs#L73) and is working as intended as well.
So I think it's safe to revert that change.

bdach replied:
I was pretty sure these calls were scheduled correctly already which is like 95% of the reason why I was confused.

Yes, please revert the change.

Susko3 reviewed (changes requested):
Test failures are relevant. `Host.Window` is null on headless tests.

zihadmahiuddin committed:
Update window title in input thread

zihadmahiuddin committed:
Skip updating window title in headless mode

bdach commented:
Not sure about the following text being in the titlebar while the intro is playing out:

![1741183072](https://github.com/user-attachments/assets/87510af9-3d7c-4af2-9b18-468c83b82a17)

Feels kinda weird and is exposing our implementation details of "the intro music is a beatmap actually". Maybe this should be done in some other way to prevent this sort of stuff. Possibly hooked up to the [`UserActivity`](https://github.com/ppy/osu/blob/f6cf63edaea5f719b1ce177c8da0dae52e4d2647/osu.Game/Users/UserActivity.cs#L37) system that we use for online and presence and so on?

Need @peppy or @smoogipoo opinion on this.

bdach commented on osu.Game/OsuGame.cs:
```diff
@@ -828,6 +828,16 @@ private void beatmapChanged(ValueChangedEvent<WorkingBeatmap> beatmap)
+            if (Host.Window == null)
+                return;
+
+            // prevent weird window title saying please load a beatmap
```
This comment can be removed I think, the `DummyWorkingBeatmap` check is self-explanatory / self-documenting and does not require elaboration IMO.

zihadmahiuddin committed:
Revert "Update window title in input thread"

This reverts commit 14b5c0bf10389b924fa8ca515c2e27457fdcc119.
This is not necessary as the title update is already scheduled on the correct thread by the framework.

zihadmahiuddin committed:
Remove unnecessary comment

bdach commented:
> I'm assuming all the "special" maps have their `BeatmapSetInfo.Protected` set to true

I forgot that this exists to be honest. We can use that too if it works, probably simpler that way.

zihadmahiuddin committed:
Skip updating window title for protected mapsets

peppy commented:
Should only show during gameplay like stable.

zihadmahiuddin committed:
Update window title changes to match osu! stable

It shows beatmap metadata during gameplay, spectating, and watching replays but shows beatmap filename during editng

smoogipoo reviewed (dismissed):
Lgtm

peppy committed:
Simplify code and don't set title if already correct
